HSEARCH-2818 Optimise encoding of GSON elements into HTTP client buffers

Sanne opened the pull request:
Similar optimisations as proposed before but using a different approach, now more carefully taking into account the AWS digest requests.

Also taking into account flow control of the underlying buffers, this was a bug in the previous proposal.

https://hibernate.atlassian.net/browse/HSEARCH-2818

yrodiere commented on elasticsearch/src/test/java/org/hibernate/search/elasticsearch/test/GsonStreamedEncodingTest.java:
```diff
@@ -0,0 +1,326 @@
+	}
+
+	@Test
+	public void testDigestToTriggerLenghtComputation() {
```
typo here, should be `testDigestToTriggerLengthComputation`

yrodiere commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/DigestSelfSigningCapable.java:
```diff
@@ -0,0 +1,22 @@
+ *
+ * @author Sanne Grinovero (C) 2017 Red Hat Inc.
+ */
+public interface DigestSelfSigningCapable {
```
See also my comment on `GsonHttpEntity#writeTo`: we may not need this interface at all.

Sanne replied:
having the explicit hint that this class is being invoked for the purpose of the digest computation helps a lot in making it efficient..

Could make it SPI but IMO it's borderline as something we'd not support others to use / rely on.
Is any invocation "across _our_ modules" needing to be an SPI ?

Sanne replied:
I've tried such things before. It didn't work out for various reasons.

1. the `writeTo(Outpustream)` is invoked in a different context by the http client and you can't do flow control in that case - not without buffering which is what I'm working to avoid.

> We could use a wrapper around the OutputStream to detect the content length if necessary.

That's too late. It's going to need the content lenght before the content is being produced, as the content lenght needs to be written in the output stream first thing, and also it might affect how the rest of the content is actually encoded (chunked, buffered, etc..).

Also you can't report a different lenght after you've produced the content as that gets you in trouble (bug of the client?).

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -0,0 +1,335 @@
+		private void writeOrBufferWrite(ByteBuffer toWriteBytes) throws IOException {
+			totalWrittenBytes += toWriteBytes.remaining();
+			flushPendingBuffers();
+			if ( flowControlPushingBack == false ) {
```
It's not as readable as an explicit `== false` :)

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -0,0 +1,335 @@
+
+	@Override
+	public void close() throws IOException {
+		//Nothing to close
```
that's not possible: the async http client expects to be able to call us multiple times and resume the data stream from the last time, so I can't just reset for the sake of it.

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -0,0 +1,335 @@
+	public void writeTo(OutputStream outstream) throws IOException {
+		//This could be implemented but would be sub-optimal compared to using produceContent().
+		//We therefore prefer throwing the exception so that we can easily spot unintended usage via tests.
+		throw new UnsupportedOperationException( "Not implemented! Expected to produce content only over produceContent()" );
```
I'll not do it as the code is already very complex, and this method isn't even ever called so not worth the effort..

Sanne replied:
BTW I don't want to buffer w/o boundarier. I'm using a buffer of a very specific size which I know is a good balance for our use case.

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -0,0 +1,335 @@
+	/**
+	 * We don't want to compute the length in advance as it would defeat all optimisations.
+	 * Still it's possible that we happen to find out, for example if a Digest from all
+	 * content needs to be computed.
```
This was actually buggy and I figured it out later: once the content length is provided we need to stick to it or the http client driver gets seriously confused. The new design is better and guarantees a "stable" response w/o drawbacks.

Sanne commented:
I've updated it, fixed the typos and applied several more optimisations.

The previous version would avoid all allocations "out of TLAB" - which are the most expensive.
After the second round, we don't see any allocations within TLAB either - the only remaining significant allocations are performed during response parsing.

yrodiere commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -192,81 +235,133 @@ public void fillDigest(MessageDigest digest) throws IOException {
+			}
+			//The original CharBuffer is not safe as it's backed by a char array
+			//which happens to be the GSON write buffer: it will get mutated.
+			CharBuffer defensiveCopy = CharBuffer.allocate( toWriteLater.remaining() );
```
It may be silly, but... since you have to perform an allocation anyway, why not directly allocate ByteBuffers, and directly encode the characters in there? You could then remove this list of char buffers, and use a list of ByteBuffers instead. This would mean turning "needsWritingBuffer" into a list, if you will: it would initially only contain one buffer, and would grow as necessary.

I tested it a bit with the performance tests, and it seems having a lot of "unwritten lazy buffers" is not rare; I was above the 16 buffer initial capacity about 25% of the time, I think, going all the way up to more than 50 buffers. So it could be useful.
But more importantly, I wonder if it couldn't make the code a bit simpler. Do you think it could?

Sanne replied:
I'm not sure I understand what you'd hope to achieve. I'd love to simplify the code a bit but not at cost of performance, and I'm not sure if you've considered all aspect .. for example that you actually don't know how large the bytebuffers need to be as that depends on the encoding details, and guessing gets you into re-sizing trouble.

I need to wrap this up, let's apply this as-is but then if you want to play with it and show me I'm happy to review some follow-up polishing PR.

yrodiere commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -192,81 +235,133 @@ public void fillDigest(MessageDigest digest) throws IOException {
+		hintContentLenght( digestWriter.getContentLenght() );
+	}
+
+	private void hintContentLenght(long contentLenght) {
```
Typo: "hintContentLenght" => "hintContentLength", "contentLenght" => "contentLength"

Sanne commented on elasticsearch/src/main/java/org/hibernate/search/elasticsearch/util/impl/GsonHttpEntity.java:
```diff
@@ -262,6 +316,15 @@ private void hintContentLenght(long contentLenght) {
+		//can calculate the accurate size upfront.
+		//Computing this size after multiple buffers is possible but not helpful,
+		//as we'll already be using the chunked encoding mode.
+		long totalWrittenBytes = 0;
```
thanks, I'll remove it. It's just a leftover from a previous approach which didn't work out.

Sanne commented:
I've rebased it and fixed the problems with missing `static final` and what you highlighted about close() not guaranteeing to make it repeatable.

Also promoted that disgest interface to SPI as suggested.

I didn't change the essential strategy around buffers as I'm not seeing a clearly better idea. Happy to be proven wrong but let's split that out as a different PR if you think it's worth trying.

yrodiere commented:
Merged, thanks! Good to see this one in :)
I'll try to submit a follow-up PR to show what I was talking about.
